ast/ast.go

```diff
@@ -1376,22 +1377,56 @@ type ArrayLiteral struct {
 	Values []Expr
 }
 
-// StructLiteral is struct literal node.
+// TupleStructLiteral is tuple syntax struct literal node.
 //
-//	STRUCT{{if not (isnil .Fields)}}<{{.Fields | sqlJoin ","}}>{{end}}({{.Values | sqlJoin ","}})
-type StructLiteral struct {
-	// pos = Struct || Lparen
+//	({{.Values | sqlJoin ","}})
+type TupleStructLiteral struct {
+	// pos = Lparen
 	// end = Rparen + 1
 
-	Struct         token.Pos // position of "STRUCT"
 	Lparen, Rparen token.Pos // position of "(" and ")"
 
-	// NOTE: Distinguish nil from len(Fields) == 0 case.
-	//       nil means type is not specified, or empty slice means this struct has 0 fields.
+	Values []Expr // len(Values) > 1
+}
+
+// TypedStructLiteral is typed struct literal node.
+//
+//	STRUCT<{{.Fields | sqlJoin ","}}>({{.Values | sqlJoin ","}})
+type TypedStructLiteral struct {
+	// pos = Struct
+	// end = Rparen + 1
+
+	Struct token.Pos // position of "STRUCT"
+	Rparen token.Pos // position of ")"
+
 	Fields []*StructField
 	Values []Expr
 }
 
+// TypelessStructLiteral is typeless struct literal node.
+//
+//	STRUCT({{.Values | sqlJoin ","}})
+type TypelessStructLiteral struct {
+	// pos = Struct
+	// end = Rparen + 1
+
+	Struct token.Pos // position of "STRUCT"
+	Rparen token.Pos // position of ")"
+
+	Values []*ExprAsName
+}
+
+// ExprAsName is value with optional name in typeless struct literal or new constructor.
```

Comment:
I think it is not only for typeless struct literal and new constructor.
We can encourage to use this struct for any `expr [AS ident]` syntax, without any semantic implications.

Reply:
We can update the document anytime.

Reply:
It means that this struct is added only for `NewConstructorArg` and `TypelessStructValue` in this PR context and we need to discuss to change this struct semantics in another PR.

Reply:
Yes.

Reply:
From this conversation, it sounds like you don't intend to widely recommend the use of this struct.

Reply:
My thought:
* `TypelessStructValue` and `NewConstructorArg` is not same abstraction, but have common concrete structure.
* Concrete structure will able to be reused without modification.
* IMO, currently `ExprAsName` is no strong reason to be introduced.